src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -505,56 +524,52 @@ private void onMqttSubscribe(
                     }
-
-                    int reasonCode = 0x8f;
                     final String topicFilter = subscription.topicFilter().asString();
                     final RouteFW route = resolveTarget(routeId, authorization, topicFilter);
 
                     if (route != null)
                     {
-                        final MqttServerStream serverStream = new MqttServerStream();
-
                         final long newRouteId = route.correlationId();
                         final long newInitialId = supplyInitialId.applyAsLong(newRouteId);
                         final long newReplyId = supplyReplyId.applyAsLong(newInitialId);
 
                         final MessageConsumer newTarget = router.supplyReceiver(newInitialId);
 
-                        final MqttBeginExFW beginEx = mqttBeginExRW
-                            .wrap(writeBuffer, BeginFW.FIELD_OFFSET_EXTENSION, writeBuffer.capacity())
-                            .typeId(mqttTypeId)
-                            .role(r -> r.set(MqttRole.RECEIVER))
-                            .clientId("client")
-                            .topic(topicFilter)
-                            .subscriptionId(subscriptionId)
-                            .build();
+                        final MqttServerStream serverStream = new MqttServerStream(newTarget,
+                            newRouteId, newInitialId, newReplyId, ackIndex, packetId);
 
-                        serverStream.doMqttBeginEx(newTarget, newRouteId, newInitialId,
-                            decodeTraceId, beginEx.buffer(), beginEx.offset(), beginEx.sizeof());
+                        serverStream.doMqttBeginEx(decodeTraceId, topicFilter, subscriptionId);
 
                         correlations.put(newReplyId, serverStream);
 
-                        reasonCode = 0x00;
-
-                        mqttStreams.put(topicFilter, serverStream);
+                        subscribers.put(topicFilter, serverStream);
+                        reasonCodeCount++;
+                    }
+                    else
+                    {
+                        unrouteableMask |= 1 << ackIndex;
                     }
-
-                    reasonCodes.add((byte) reasonCode);
                 }
             }
-            byte[] subackReasonCodes = new byte[reasonCodes.size()];
-            for (int i = 0; i < reasonCodes.size(); i++)
+
+            Subscription subscription = new Subscription(unrouteableMask, reasonCodeCount);
+            subscriptionsByPacketId.put(packetId, subscription);
+
+            for (int i = 0; i < reasonCodeCount; i++)
             {
-                subackReasonCodes[i] = reasonCodes.get(i);
+                if ((unrouteableMask & (1 << i)) > 0)
+                {
+                    subscription.onSubscribeFailed(packetId, i, 0x8f);
+                }
             }
-            doMqttSuback(subackReasonCodes);
+            subscribers.forEach((filter, stream) -> stream.subscription = subscription);
```

Comment:
Prefer to pre-create the `Subscription`, pass to each `MqttServerStream`, then after all streams processed, update `subscription.ackCount` to `reasonCodeCount` and `subscription.ackMask` to `unrouteableMask`.